### PR TITLE
BUG: Handle zero-chunked pyarrow.ChunkedArray in StringArray

### LINKED ISSUE #41040: BUG: Cannot construct StringArray from pyarrow.ChunkedArray with zero chunks
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pyarrow as pa
import pandas as pd

df = pd.DataFrame({"string": pd.array([], dtype="string")})
table = pa.Table.from_pandas(df)
print(table.column(0))
# <pyarrow.lib.ChunkedArray object at 0x12c5cec70>
# [
#   []
# ]

# Construct a Table with zero chunks as returned by `pyarrow.dataset`
empty_table = pa.table([pa.chunked_array([], type=pa.string())], schema=table.schema)
print(empty_table.column(0))
# <pyarrow.lib.ChunkedArray object at 0x12c5cee00>
# [
# 
# ]

empty_table.to_pandas()
```

yields the following traceback:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-19-5047304e95a2> in <module>
----> 1 empty_table.to_pandas()

~/mambaforge/envs/fletcher/lib/python3.8/site-packages/pyarrow/array.pxi in pyarrow.lib._PandasConvertible.to_pandas()

~/mambaforge/envs/fletcher/lib/python3.8/site-packages/pyarrow/table.pxi in pyarrow.lib.Table._to_pandas()

~/mambaforge/envs/fletcher/lib/python3.8/site-packages/pyarrow/pandas_compat.py in table_to_blockmanager(options, table, categories, ignore_metadata, types_mapper)
    790     _check_data_column_metadata_consistency(all_columns)
    791     columns = _deserialize_column_index(table, all_columns, column_indexes)
--> 792     blocks = _table_to_blocks(options, table, categories, ext_columns_dtypes)
    793 
    794     axes = [columns, index]

~/mambaforge/envs/fletcher/lib/python3.8/site-packages/pyarrow/pandas_compat.py in _table_to_blocks(options, block_table, categories, extension_columns)
   1131     result = pa.lib.table_to_blocks(options, block_table, categories,
   1132                                     list(extension_columns.keys()))
-> 1133     return [_reconstruct_block(item, columns, extension_columns)
   1134             for item in result]
   1135 

~/mambaforge/envs/fletcher/lib/python3.8/site-packages/pyarrow/pandas_compat.py in <listcomp>(.0)
   1131     result = pa.lib.table_to_blocks(options, block_table, categories,
   1132                                     list(extension_columns.keys()))
-> 1133     return [_reconstruct_block(item, columns, extension_columns)
   1134             for item in result]
   1135 

~/mambaforge/envs/fletcher/lib/python3.8/site-packages/pyarrow/pandas_compat.py in _reconstruct_block(item, columns, extension_columns)
    749             raise ValueError("This column does not support to be converted "
    750                              "to a pandas ExtensionArray")
--> 751         pd_ext_arr = pandas_dtype.__from_arrow__(arr)
    752         block = _int.make_block(pd_ext_arr, placement=placement,
    753                                 klass=_int.ExtensionBlock)

~/Development/pandas/pandas/core/arrays/string_.py in __from_arrow__(self, array)
    119             results.append(str_arr)
    120 
--> 121         return StringArray._concat_same_type(results)
    122 
    123 

~/Development/pandas/pandas/core/arrays/_mixins.py in _concat_same_type(cls, to_concat, axis)
    240         dtypes = {str(x.dtype) for x in to_concat}
    241         if len(dtypes) != 1:
--> 242             raise ValueError("to_concat must have the same dtype (tz)", dtypes)
    243 
    244         new_values = [x._ndarray for x in to_concat]

ValueError: ('to_concat must have the same dtype (tz)', set())
```

#### Problem description

The problem here is that we are checking in `_concat_same_type` whether we concat arrays of the same type but as we don't concatenate anything, there are also no types to check. This should instead create an empty array.

### PR DESCRIPTION
**Pull request by xhochy**
- [x] closes #41040
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/arrays/boolean.py [hunk 1]
```diff
@@ -126,7 +126,12 @@ def __from_arrow__(
             bool_arr = BooleanArray._from_sequence(np.array(arr))
             results.append(bool_arr)
 
-        return BooleanArray._concat_same_type(results)
+        if len(results) == 0:
```
Comment (phofl) [inline pandas/core/arrays/boolean.py:129]: not results here instead of length check?

Comment (xhochy) [reply pandas/core/arrays/boolean.py:129]: What would be the reasoning for this? `== 0` is the more readable version IMO.

**Comment by xhochy** [comment]
The test failure on Windows is unrelated and also happening on Windows.

File: pandas/core/arrays/boolean.py (hunk 1 above)
Comment (phofl) [reply pandas/core/arrays/boolean.py:129]: Consistency, I think we are mostly using not result in these cases, but might be wrong here

**Comment by phofl** [comment]
This happens sometimes, ``ResourceWarning`` yes

File: pandas/core/arrays/boolean.py (hunk 1 above)
Comment (xhochy) [reply pandas/core/arrays/boolean.py]: Updated and CI passed except for the `ResourceWarning`.

**Review by phofl — APPROVED** [review]

File: doc/source/whatsnew/v1.3.0.rst [hunk 2]
```diff
@@ -700,7 +700,7 @@ Conversion
 Strings
 ^^^^^^^
 
--
+- Bug in the conversion from ``pyarrow.ChunkedArray`` to :class:`StringArray` when the original had zero chunks (:issue:`41040`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.3.0.rst:703]:
```suggestion
- Bug in the conversion from ``pyarrow.ChunkedArray`` to :class:`~arrays.StringArray` when the original had zero chunks (:issue:`41040`)
```

File: pandas/core/arrays/boolean.py [hunk 3]
```diff
@@ -126,7 +126,12 @@ def __from_arrow__(
             bool_arr = BooleanArray._from_sequence(np.array(arr))
             results.append(bool_arr)
 
-        return BooleanArray._concat_same_type(results)
+        if not results:
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/boolean.py:129]:
```suggestion
        if len(results) == 0:
```

? (I find that more explicit / readable)

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for working on this! 

Period and Interval might have the same problem (but also fine to leave it as follow-up)

I was kind of wondering if `_concat_same_type` should be able to handle the "empty list of arrays" case. But in the current implementation (as a class method on the EA, without access to the actual `dtype` instance), that's not possible, because it wouldn't know the dtype for the resulting array.

File: pandas/core/arrays/boolean.py (hunk 3 above)
Comment (xhochy) [reply pandas/core/arrays/boolean.py:129]: I can simply revert https://github.com/pandas-dev/pandas/pull/41052/commits/f4a2c8c90b5e8c0218869c5eff30354603b3bca5 (did do that commit as @phofl suggested)

Comment (jorisvandenbossche) [reply pandas/core/arrays/boolean.py:129]: Whoops, missed that comment. Sorry for the contradicting feedback ;)

(let's leave it for now then, something to discuss / agree on as pandas team)

**Comment by xhochy** [comment]
> Period and Interval might have the same problem (but also fine to leave it as follow-up)

I'll fix that here, too.

File: pandas/core/arrays/boolean.py (hunk 3 above)
Comment (phofl) [reply pandas/core/arrays/boolean.py:129]: Saw this on a few prs in the past that either not was used or it was recommended to use not, hence my feedback, sorry if I misinterpreted this

**Review by jorisvandenbossche — APPROVED** [review]

**Pull request merged**